src/dali/plugin/input/rest/kucoin_futures.py

```diff
@@ -0,0 +1,288 @@
+
+    @property
+    def _client(self) -> kucoinfutures:
+        super_client: Exchange = super()._client
+        if not isinstance(super_client, kucoinfutures):
+            raise TypeError("super_client is not of type kucoinfutures")
+
+        return super_client
+
+    # using kucoin transaction history api instead. (implicit api)
+    def _get_process_deposits_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_withdrawals_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_trades_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _process_gains(
+        self,
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+    ) -> None:
+        pass
+
+    def _fetch_data(
+        self,
+        start_at: int,
+        end_at: int,
+    ) -> List[Dict[str, Any]]:
+
+        offset: int = 1
+        retries: int = 0
+        has_more: bool = False
+
+        ledger_transactions: Dict[str, Any] = {}
+        items: List[Dict[str, Any]] = []
+
+        while (start_at <= end_at) and retries < 4:
+            try:
+                ledger_transactions = self._client.futuresPrivateGetTransactionHistory({_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT})
+                data_list = ledger_transactions[_DATA][_DATA_LIST]
+                has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                for item in data_list:
+                    items.append(item)
+                    offset = item[_OFFSET]
+
+                # https://github.com/ccxt/ccxt/issues/10273
+                # enableRateLimit is not working for kucoin
+                time.sleep(0.15)
+                retries = 0
+
+                while has_more and (retries < 4):
+                    try:
+                        ledger_transactions = self._client.futuresPrivateGetTransactionHistory(
+                            {_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT, _OFFSET: offset}
+                        )
+                        data_list = ledger_transactions[_DATA][_DATA_LIST]
+                        has_more = ledger_transactions[_DATA][_HAS_MORE]
+
+                        for item in data_list:
+                            items.append(item)
+                            offset = item[_OFFSET]
+
+                        time.sleep(0.15)
+
+                    except RateLimitExceeded:
+                        self._client.sleep(13000)
+                        retries += 1
+
+                if retries >= 4:
+                    raise Exception("Failed to fetch ledger transactions after 4 retries")
+
+                start_at += _TS_INCREMENT
+                offset = 1
+                has_more = False
+                retries = 0
+
+            except RateLimitExceeded:
+                self._client.sleep(13000)
```

Comment:
Is this a multi-hour long blocking sleep? If so we should just raise an exception instead.

Reply:
I should've added a comment line. sleep takes time in milliseconds. So, it basically pauses for 13 seconds and retries.
Added comment.